src/tail/mod.rs

```diff
@@ -52,28 +53,19 @@ impl Tail {
             let tunnel = tokio::spawn(tunnel_process.run(tunnel_rx));
 
             // Register the tail with the Workers API and send periodic heartbeats
-            let session = tokio::spawn(Session::run(target, user, session_rx));
+            let session = tokio::spawn(Session::run(target, user, session_rx, tx));
 
-            let res = tokio::try_join!(listener, log_server, session, tunnel);
+            let res = tokio::try_join!(
+                async { listener.await? },
+                async { log_server.await? },
+                async { session.await? },
+                async { tunnel.await? }
+            );
 
             match res {
                 Ok(_) => Ok(()),
-                Err(e) => failure::bail!(e),
+                Err(e) => Err(e),
```

Comment:
this is interesting

Reply:
i'm not 100% certain this was necessary. there was a lot of shotgunning on this one.